src/internal/consistenthashing/etcd.go

```diff
@@ -0,0 +1,291 @@
+// until it is shutdown.
+type node struct {
+	Member
+	Locks map[string]lockInfo
+}
+
+// Member contains metadata about members in a Ring.
+type Member struct {
+	Id   string
+	Hash uint32
+}
+
+// lockInfo instances are created for keys by a ring's node.
+// A context lockInfo.Ctx is stored so the proper context is used when unlocking lockInfo.Lock
+type lockInfo struct {
+	Lock dlock.DLock
+	Ctx  context.Context
+}
+
+// New is the default constructor for Ring. Creating a Ring starts a goroutine to watch prefix in etcd and creates a
+// node that continuously refreshes a lease in etcd.
+func New(ctx context.Context, client *etcd.Client, logger *logrus.Logger, prefix string, id ...string) (*Ring, error) {
+	if id == nil {
+		id = []string{uuid.New()}
+	}
+	localMember := Member{
+		Id:   id[0],
+		Hash: crc32.ChecksumIEEE([]byte(id[0])),
+	}
+	ring := &Ring{
+		client:  client,
+		hashFn:  crc32.ChecksumIEEE,
+		logger:  logger,
+		members: []Member{localMember},
+		node: node{
+			Member: localMember,
+			Locks:  map[string]lockInfo{},
+		},
+		prefix:    prefix,
+		stateLock: sync.Mutex{},
+	}
+	etcdCol := collection.NewEtcdCollection(ring.client, path.Join(ring.prefix, "nodes"),
+		nil, &types.BoolValue{}, nil, nil)
+	go ring.watch(ctx, etcdCol)
+	go ring.createNode(ctx, etcdCol, id[0])
+	return ring, nil
+}
+
+// GetIndex returns the index of the Member associated with key. A key belongs to the first Member whose hash is
+// greater than the key. If no Members are greater, then the Ring wraps around to the first Member.
+func (ring *Ring) GetIndex(key string) int {
+	keyHash := ring.hashFn([]byte(key))
+	index := sort.Search(len(ring.members), func(i int) bool {
+		return ring.members[i].Hash >= keyHash
+	})
+	// wrap around to the first node in the Ring.
+	if index >= len(ring.members) {
+		index = 0
+	}
+	return index
+}
+
+// Get is a convenience wrapper around GetIndex that returns the Member that associates to a key.
+func (ring *Ring) Get(key string) Member {
+	return ring.members[ring.GetIndex(key)]
+}
+
+// Lock attempts to lock a key in etcd using consistent hashing. If key does not belong to the ring's node,
+// the call blocks until context is cancelled.
+func (ring *Ring) Lock(ctx context.Context, key string) (context.Context, error) {
+	ring.stateLock.Lock()
+	defer ring.stateLock.Unlock()
+	ticker := time.NewTicker(time.Second * 5)
+	defer ticker.Stop()
+	// todo: this should be event driven instead of by polling for better scalability.
+	for {
+		if len(ring.members) != 0 {
+			index := ring.GetIndex(key)
+			if ring.node.Id == ring.members[index].Id {
+				break
+			}
+		}
+		ring.stateLock.Unlock()
+		select {
+		case <-ctx.Done():
+			ring.stateLock.Lock()
+			return nil, errors.EnsureStack(ctx.Err())
+		case <-ticker.C:
+		}
+		ring.stateLock.Lock()
+	}
+	if info, alreadyExists := ring.node.Locks[key]; alreadyExists {
+		lockCtx := info.Ctx
+		return lockCtx, nil
+	}
+	nodeLock := dlock.NewDLock(ring.client, key)
```

Comment:
Should the prefix be a part of this?

Reply:
I'm not sure. If it is, then two rings could lock the same key:
```go
ring1.Lock(ctx, "repo1")
ring2.Lock(ctx, "repo2")
``` 
where the lock creation would look like this
```
nodeLock := dlock.NewDLock(ring.client, ring1.Prefix + "/" + key)
nodeLock := dlock.NewDLock(ring.client, ring2.Prefix + "/" + key)
```
I don't think we want to be able to do that right?